Add worker tests

kitcambridge opened the pull request:
This pull request adds tests for the worker functions, and @jrconlin's worker benchmarks. Apart from the tests, most of the additions are mocks for the new `Socket`, `Server`, and `Handler` interfaces. Some of the worker methods have also been refactored for testability.

* The `Socket` interface is now used by `WorkerWS` for interacting with the WebSocket. There's a `WebSocket` implementation that wraps `websocket.Conn` and a `MockSocket`.
* `WorkerWS.sniffer()` now uses a read deadline instead of `time.AfterFunc()` for the client handshake timeout. This avoids spawning an extra goroutine per connection, and makes it easier to test the worker run loop. The `pongInterval` is used as the deadline after the client sends a `hello` packet.
* `Server` is now an interface, making it easier to test the interactions between the worker and parent server.
* The client smoke tests now live in separate files, with a `smoke` build tag. This makes it easier to separate them out from the server unit tests.

Closes #183. Part of #132.

@bbangert, @jrconlin r?

kitcambridge commented on src/github.com/mozilla-services/pushgo/simplepush/structs.go:
```diff
@@ -42,8 +40,8 @@ type PushCommand struct {
 
 type PushWS struct {
 	uaidLock sync.RWMutex
```
Noting for the future: now that we're using a deadline instead of `time.AfterFunc` for the handshake timeout, I think we can remove the mutex around `uaid`.

jrconlin commented on src/github.com/mozilla-services/pushgo/simplepush/config_test.go:
```diff
@@ -94,9 +94,15 @@ func TestConfigFile(t *testing.T) {
-		t.Errorf("Mismatched origins: got %#v; want %#v",
-			app.sh.origins, origins)
+	socketHandler := app.SocketHandler()
+	if sh, ok := socketHandler.(*SocketHandler); ok {
```
I completely understand why and the approach, but, yeah, all the alarm bells are ringing in my head about this too. Unit tests are fun ways to remove type safety in Go.

jrconlin commented on src/github.com/mozilla-services/pushgo/simplepush/worker.go:
```diff
@@ -713,10 +720,11 @@ func (self *WorkerWS) Flush(sock *PushWS, lastAccessed int64, channel string, ve
+			logStrings[i] = fmt.Sprintf("%s %s.%s = %d", prefix, uaid,
+				update.ChannelID, update.Version)
 		}
+		self.metrics.IncrementBy("updates.sent", int64(len(updates)))
```
Good catch!

jrconlin commented:
r+

kitcambridge commented on src/github.com/mozilla-services/pushgo/simplepush/config_test.go:
```diff
@@ -134,9 +140,14 @@ func TestConfigFile(t *testing.T) {
-	if handlers.maxDataLen != 512 {
-		t.Errorf("Wrong maximum data size: got %d; want 512", handlers.maxDataLen)
+	endpointHandler := app.EndpointHandler()
+	if eh, ok := endpointHandler.(*EndpointHandler); ok {
```
![Eh?](https://cloud.githubusercontent.com/assets/243830/5684595/1af7fea4-97ef-11e4-93ab-97bb5904b0f8.jpg)

jrconlin replied:
/me waits for either the MuxEndpointHandler variable (which I'm kinda ambivalent about) or the Functional Utility CallbacK (which might require an HR button press).

bbangert commented on src/github.com/mozilla-services/pushgo/simplepush/simplepush.go:
```diff
@@ -0,0 +1,43 @@
+
+func init() {
+	revertMocks()
+}
```
I don't really understand the naming of this file nor its purpose. There are no doc strings in it, and init() always runs, so revertMocks will always run.

jrconlin replied:
I let this slide. It's setting up things for testing. For what it's worth, I found "revertMocks" to be a bit odd, particularly since it's the norm. I think we can Kit could rename this as testwrapper.go and avoid some of the unit test file name issues.

kitcambridge replied:
Ah, yes; the file of shame. I'll rename to `mock_test.go` and add some comments. These were only used in `log_emitters_test.go` (specifically for testing log message generation), but came in handy for the worker tests, too.

 `installMocks` replaces `id.Generate{Bytes}`, `os.Getpid`, and `time.Now` with deterministic functions; `revertMocks` restores them. Instead of duplicating the assignments in the `var` block and `revertMocks`, I just opted for empty variable declarations and a `revertMocks` call on `init`.

kitcambridge replied:
Here's how I had it originally:

```go
var (
	idGenerate = id.Generate
	idGenerateBytes = id.GenerateBytes
	osGetPid = os.Getpid
	timeNow = time.Now
)

func revertMocks() {
	idGenerate = id.Generate
	idGenerateBytes = id.GenerateBytes
	osGetPid = os.Getpid
	timeNow = time.Now
}
```

Since `revertMocks` sets the defaults anyway, I used `init()` to avoid repeating the assignments. Is the original way clearer?

bbangert commented on src/github.com/mozilla-services/pushgo/simplepush/socket.go:
```diff
@@ -0,0 +1,157 @@
+	"golang.org/x/net/websocket"
+)
+
+type Socket interface {
```
Doc strings? This is all for what purpose? I can guess its for mocking capability, but it'd be nice to doc these things so we remember in the future.

bbangert commented on src/github.com/mozilla-services/pushgo/simplepush/socket.go:
```diff
@@ -0,0 +1,157 @@
+	}
+}
+
+type RecorderSocket struct {
```
If this is for testing, why not use a mock one generated off the interface?

kitcambridge replied:
Good catch! `MockSocket` and `gomock.InOrder` is more than adequate.

bbangert commented on src/github.com/mozilla-services/pushgo/simplepush/worker.go:
```diff
@@ -50,8 +47,9 @@ type WorkerWS struct {
-	WorkerActive               = 1
+	WorkerInactive WorkerState = iota
+	WorkerActive
+	WorkerStopped
```
Nice cleanup!

bbangert commented on src/github.com/mozilla-services/pushgo/simplepush/worker.go:
```diff
@@ -136,35 +140,47 @@ func harmlessConnectionError(err error) (harmless bool) {
+	}
+	if d > 0 {
+		t = timeNow().Add(d)
+	}
```
If d is not > 0, then t remains uninitialized? That's ok?

kitcambridge replied:
Yup; in that case, no deadline will be set. Good candidate for a doc string. :grinning:

bbangert commented on src/github.com/mozilla-services/pushgo/simplepush/worker.go:
```diff
@@ -789,5 +801,13 @@ func (r *NoWorker) Flush(_ *PushWS, lastAccessed int64, channel string, version
 	return nil
 }
 
+func cleanChannelIDs(ids []json.RawMessage) (clean []string) {
```
This appears to remove the front and last character of every item. The function name isn't terribly descriptive, nor is it clear why they need to be 'cleaned'. A doc string would help a lot.

bbangert commented:
r- pending changes

kitcambridge commented:
@bbangert Thanks for reviewing! Addressed your comments:

* Renamed `{install, revert}Mocks` to `use{Mock, Std}Funcs`, respectively.
* Moved `useMockFuncs` into `mock_funcs_test.go`, so they're never included in the binaries.
* Added docs.
* Replaced all uses of `RecorderSocket` with `MockSocket`, and removed the former.
* Removed `cleanChannelIDs` entirely.

bbangert commented:
r+
